src/debugger/StopEventManager.js

```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import invariant from "./../invariant.js";
+import type { DebugChannel } from "./channel/DebugChannel.js";
+import { Breakpoint } from "./Breakpoint.js";
+import { Stepper, StepIntoStepper, StepOverStepper } from "./Stepper.js";
+import { BabelNode } from "babel-types";
+
+export type StoppableObject = Breakpoint | Stepper;
+
+export class StopEventManager {
+  constructor(channel: DebugChannel) {
+    this._channel = channel;
+  }
+  _channel: DebugChannel;
+
+  shouldDebuggeeStop(ast: BabelNode, stoppables: Array<StoppableObject>): boolean {
```

Comment:
It seems that this returns true if stoppables contains at least one element. It also seems that only the first element is actually used. This seems highly surprising to me.

Reply:
Stoppables is list of objects that the debuggee should stop on (breakpoints, completed steppers). Having at least one element means there is at least one reason for the debuggee to stop on this ast node. When the debuggee should stop, a reason is sent to the UI. When there are multiple reasons to stop, only one reason is needed. I currently choose this to be the reason of the first stoppable. I aggregate all the possible stoppables here so we can potentially do some more sophisticated logic to choose which reason to send.

Reply:
Make this as comment for this method.

Reply:
In that case the name of the method is highly confusing and it makes little sense for it to return a value.

Reply:
I have changed the name to `getDebuggeeStopReason` and have it return the reason or void. I have also moved the channel call outside.